oauth/servlet.go

```diff
@@ -0,0 +1,155 @@
+package oauth
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+
+	"github.com/Shopify/goose/logger"
+	"github.com/Shopify/goose/srvutil"
+)
+
+var (
+	ErrUnauthenticated = errors.New("user is not authenticated")
+	ErrLoginDisabled   = errors.New("login disabled")
+)
+
+type Servlet interface {
+	srvutil.Servlet
+	Middleware(next http.Handler) http.Handler
+}
+
+func NewServlet(manager Manager, paths *Paths) Servlet {
+	return &servlet{
+		manager: manager,
+		paths:   paths,
+	}
+}
+
+type servlet struct {
+	manager Manager
+	paths   *Paths
+}
+
+func (s *servlet) RegisterRouting(r *mux.Router) {
+	r.HandleFunc(s.paths.LoginPath, s.loginHandler)
+	r.HandleFunc(s.paths.CallbackPath, s.callbackHandler)
+}
+
+func (s *servlet) loginEnabled(w http.ResponseWriter, r *http.Request) bool {
+	if s.manager != nil {
+		return true
+	}
+
+	log(r.Context(), nil).Error("login not configured")
+	http.Error(w, "Unable to initiate login.", http.StatusInternalServerError)
+	return false
+}
+
+func (s *servlet) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, err := s.authorize(w, r)
+		if err != nil {
+			log(r.Context(), err).Warn("user is not authorized")
+			redirect(r.Context(), w, r, s.paths.LoginURL(r.URL.Path).String())
+			return
+		}
+		r = r.WithContext(logger.WithLoggable(r.Context(), user))
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (s *servlet) authorize(w http.ResponseWriter, r *http.Request) (*User, error) {
+	ctx := r.Context()
+
+	if !s.loginEnabled(w, r) {
+		return nil, ErrLoginDisabled
+	}
+
+	token, ok := s.manager.GetSavedToken(r)
+	if !ok {
+		return nil, ErrUnauthenticated
+	}
+
+	user, err := s.manager.AuthorizeToken(ctx, token)
```

Comment:
Isn't this rather expensive to do on every request? It'll add a some 50-100ms, or more, to every request. Shouldn't we be able to cache the result for a little while at least?

Reply:
I added a TODO, but for the purpose I was using this, it wasn't a big deal. Room for future improvement 😄